### ISSUE TITLE
Ptpython install fails

### ISSUE THREAD

**Issue by ifuchs** [holzschu/a-shell#217]
I tried installing ptpython via pip and got:

Building wheel for regex (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /private/var/containers/Bundle/Application/0D2ED635-057F-476E-B7BA-9FE1C7BA2537/a-Shell.app/Library/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/mobile/Containers/Data/Application/9A89E63C-AC7D-4079-A455-862123C66D73/tmp/pip-install-4d033nzk/regex_3902b453275544fe8b648337638336af/setup.py'"'"'; __file__='"'"'/private/var/mobile/Containers/Data/Application/9A89E63C-AC7D-4079-A455-862123C66D73/tmp/pip-install-4d033nzk/regex_3902b453275544fe8b648337638336af/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /private/var/mobile/Containers/Data/Application/9A89E63C-AC7D-4079-A455-862123C66D73/tmp/pip-wheel-xhhzsw42
       cwd: /private/var/mobile/Containers/Data/Application/9A89E63C-AC7D-4079-A455-862123C66D73/tmp/pip-install-4d033nzk/regex_3902b453275544fe8b648337638336af/
  Complete output (15 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.macosx-11.2-x86_64-3.9
  creating build/lib.macosx-11.2-x86_64-3.9/regex
  copying regex_3/__init__.py -> build/lib.macosx-11.2-x86_64-3.9/regex
  copying regex_3/regex.py -> build/lib.macosx-11.2-x86_64-3.9/regex
  copying regex_3/_regex_core.py -> build/lib.macosx-11.2-x86_64-3.9/regex
  copying regex_3/test_regex.py -> build/lib.macosx-11.2-x86_64-3.9/regex
  running build_ext
  building 'regex._regex' extension
  creating build/temp.macosx-11.2-x86_64-3.9
  creating build/temp.macosx-11.2-x86_64-3.9/regex_3
  error: ('There are no compilers available on iOS, sorry. Command was: ', ['-I/private/var/containers/Bundle/Application/0D2ED635-057F-476E-B7BA-9FE1C7BA2537/a-Shell.app/Library/include/python3.9', '-c'])
  ----------------------------------------
  ERROR: Failed building wheel for regex
  Running setup.py clean for regex
Failed to build regex
Installing collected packages: regex, pathspec, mypy-extensions, click, appdirs, prompt-toolkit, black, ptpython
    Running setup.py install for regex ... error
    ERROR: Command errored out with exit status 1:
     command: /private/var/containers/Bundle/Application/0D2ED635-057F-476E-B7BA-9FE1C7BA2537/a-Shell.app/Library/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/mobile/Containers/Data/Application/9A89E63C-AC7D-4079-A455-862123C66D73/tmp/pip-install-4d033nzk/regex_3902b453275544fe8b648337638336af/setup.py'"'"'; __file__='"'"'/private/var/mobile/Containers/Data/Application/9A89E63C-AC7D-4079-A455-862123C66D73/tmp/pip-install-4d033nzk/regex_3902b453275544fe8b648337638336af/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /private/var/mobile/Containers/Data/Application/9A89E63C-AC7D-4079-A455-862123C66D73/tmp/pip-record-fn_d8m0k/install-record.txt --single-version-externally-managed --user --prefix= --compile --install-headers /var/mobile/Containers/Data/Application/9A89E63C-AC7D-4079-A455-862123C66D73/Library/include/python3.9/regex
         cwd: /private/var/mobile/Containers/Data/Application/9A89E63C-AC7D-4079-A455-862123C66D73/tmp/pip-install-4d033nzk/regex_3902b453275544fe8b648337638336af/
    Complete output (15 lines):
    running install
    running build
    running build_py
    creating build
    creating build/lib.macosx-11.2-x86_64-3.9
    creating build/lib.macosx-11.2-x86_64-3.9/regex
    copying regex_3/__init__.py -> build/lib.macosx-11.2-x86_64-3.9/regex
    copying regex_3/regex.py -> build/lib.macosx-11.2-x86_64-3.9/regex
    copying regex_3/_regex_core.py -> build/lib.macosx-11.2-x86_64-3.9/regex
    copying regex_3/test_regex.py -> build/lib.macosx-11.2-x86_64-3.9/regex
    running build_ext
    building 'regex._regex' extension
    creating build/temp.macosx-11.2-x86_64-3.9
    creating build/temp.macosx-11.2-x86_64-3.9/regex_3
    error: ('There are no compilers available on iOS, sorry. Command was: ', ['-I/private/var/containers/Bundle/Application/0D2ED6
35-057F-476E-B7BA-9FE1C7BA2537/a-Shell.app/Library/include/python3.9', '-c'])
    ----------------------------------------
ERROR: Command errored out with exit status 1: /private/var/containers/Bundle/Application/0D2ED635-057F-476E-B7BA-9FE1C7BA2537/a-S
hell.app/Library/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/mobile/Containers/Data/Appli
cation/9A89E63C-AC7D-4079-A455-862123C66D73/tmp/pip-install-4d033nzk/regex_3902b453275544fe8b648337638336af/setup.py'"'"'; __file_
_='"'"'/private/var/mobile/Containers/Data/Application/9A89E63C-AC7D-4079-A455-862123C66D73/tmp/pip-install-4d033nzk/regex_3902b45
3275544fe8b648337638336af/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', 
'"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /private/var/mobile/Containers/Data/Applic
ation/9A89E63C-AC7D-4079-A455-862123C66D73/tmp/pip-record-fn_d8m0k/install-record.txt --single-version-externally-managed --user -
-prefix= --compile --install-headers /var/mobile/Containers/Data/Application/9A89E63C-AC7D-4079-A455-862123C66D73/Library/include/
python3.9/regex Check the logs for full command output.

**Comment by holzschu** [comment]
That's interesting. ptpython depends on regex, which has C components. I just ported regex for Carnets (it's in the TestFlight version), so it will cross over to a-Shell soon (the Python part of the two repositories is synced regularly, once I'm reasonably certain there are no bugs).

**Comment by holzschu** [comment]
With the new TestFlight build (162, currently uploading), which contains `regex`, `pip install ptpython` works, but I'm not sure `ptpython` itself works as expected.

**Comment by ifuchs** [comment]
﻿
Yes, it can now be installed and run and yes, it does not work as it should. Cursor position requests (CPR) have to be supported in the terminal for it to work.

**Comment by holzschu** [comment]
Thanks for the explanation.
Is the handling of CPR done inside ptpython itself, or inside another package?

**Comment by ifuchs** [comment]
I think that ordinarily it’s done by the terminal app in which you are
running python (e.g., iTerm on the Mac). I suppose that ashell would need
to support CPR.

**Comment by ifuchs** [comment]
If you should decide to support it, it would probably be useful in Carnets
as well.

**Comment by holzschu** [comment]
I meant: looking through the source code of ptpython, I don't see where it is calling for CPR. I was wondering if this (calling for CPR, not handling it) was done in ptpython, or in another package that it loads.

**Comment by holzschu** [comment]
Looking for another issue, I found this one, which is related to the `ptpython` issue: https://github.com/holzschu/a-shell/issues/206
Short version: Commands that are interactive, like `vim`, `less` or `ipython`, must be declared beforehand (because they handle input directly). I'm going to add `ptpython` to the list, and we'll see if that helps.

**Comment by holzschu** [comment]
`ptpython` now seems to work with build 163 (currently on TestFlight). I'll let you check whether the full behaviour is present.
I should probably edit `prompt-toolkit` so that any python package that uses it becomes interactive (like `jsi` or `ptpython`).

**Comment by ifuchs** [comment]
﻿
Ptpython still has some issues with the cursor (e.g., after entering a line there should be a new line with a prompt string).  Also, emacs relies on function keys which AFAIK are not available.

**Comment by holzschu** [comment]
I do see the new line with the prompt string after entering a line. It's weird that our experiences are different.
For the emacs question, there is the option of using an external keyboard... or change the ptpython preferences to use vi bindings.

**Comment by ifuchs** [comment]
Is this what you see:

**Comment by holzschu** [comment]
Here is what I have: 
![44C87BBF-C1C2-43E9-8A78-733A2FEC0744](https://user-images.githubusercontent.com/8070736/120105452-362b7580-c159-11eb-9e2d-d48e924d3a5e.png)

(your image did not upload yet)

**Comment by ifuchs** [comment]
I tried again with an external keyboard and it is better but there are still issues. Also, unction keys (e.g., F3 history) do not work properly.

**Comment by holzschu** [comment]
Yes, in my experience too, F3 for history behaves strangely. But F2 for menu and F4 to switch modes work perfectly. I suspect history tries accessing the terminal auxiliary screen.

**Comment by holzschu** [comment]
With the new build on TestFlight (165), F3 now works -- assuming you have a way to enter F3, so probably an external keyboard.

**Comment by ifuchs** [comment]
Yes, the problem I have on my iPad Pro with Magic keyboard is there does not appear to be a way to use Function keys.
